README
$Id: README,v 1.2 2005/06/18 12:36:18 kreator Exp $

About
-----
This is complete pam_skey modul as interface to existing S/Key
library/interface by Wietse Venema. So far it has been tested on Linux
(Debian potato, Debian woody and Debian sid), Solaris 2.6, 2.7 and 2.8. I
have reason to believe it will work where PAM implementation exists and is
properly done. It provides `auth' services only. It can use tokens from
PAM_AUTHTOK to implement flexibile module stacking.
NOTE: You *absolutely* need libskey to make this module working. I
recommend original logdaemon suite by Wietse Venema - you can find it in
logdaemon-5.10 package.

This suite consists of following modules:
pam_skey.so           - Main module than by default contains routines for
                        checking skey.access for access rules
pam_skey_access.so    - Add-on module that is useful only when it is
                        required (requisite) module after pam_skey, which
                        means skey.access controls future loading of other
                        modules. Note that this emulates behavior of
                        original Venema's skeylogin.

Known problems
--------------
Note that there are several problems with existing (original) Wietse
Venema's implementation - no flock() on keys file, hairy host checking
code in skeyaccess(), etc. If there is any other library function and
parametric compatible with S/Key library, this will most probably work
with it, too. I am planning to re-implement S/Key library later, having
Venema's and Linux-skey package as main resources.

Known bugs
----------
None reported so far. That does not mean there are none: if you find any
bug, or something that is likely to be bug, try to exactly see what
happens using `debug' option in local pam configuration file and e-mail me
together with console and syslog information. Or at least e-mail me with
feedback that can help to trace problems.

Installation
------------
Please please please, for installation instructions look closely at
INSTALL file.

Literature and other sources
----------------------------
I find most helpful was Linux-PAM documentation. I had also a peek into
following sources:
  Wietse Venema's logdaemon package
  Olaf Kirch's Linux S/Key package
  Wyman Miles' pam_securid module
  Linux-PAM modules and templates

Contact
-------
My e-mail address is Dinko Korunic <kreator@srce.hr>. I usually respond
e-mails within a couple of hours, perhaps one day. If necessary, use PGP
key 0xEA160D0B, with fingerprint 48BF A3C2 6FB9 14EC B05C  A683 6D3B AED1
EA16 0D0B.


ChangeLog
2001-10-11 20:03  kreator

	* configure, defs.h.in, autoconf/config.guess, autoconf/config.sub,
	autoconf/configure.in: - update defs and configure to 2.50

2001-10-06 11:50  kreator

	* RELEASENOTES: - do releasenotes.

2001-10-06 11:45  kreator

	* ChangeLog, configure, autoconf/configure.in: - fixed nsl/socket
	dlopen() issue with proftpd 1.22 on Solaris

2001-08-16 10:23  kreator

	* contrib/: README, linux-skey-wuftpd.patch, netmatch.patch,
	norb.patch, pam_skey.c: - added contrib

2001-08-16 10:22  kreator

	* ChangeLog, INSTALL, README, RELEASENOTES, TODO, configure,
	defs.h.in, misc.h.in, pam_skey.c, pam_skey.h, pam_skey_access.c,
	autoconf/acconfig.h, autoconf/configure.in, sample/pam.conf: - misc

2001-03-14 23:56  kreator

	* ChangeLog, RELEASENOTES: Complete releasenotes.

2001-03-14 23:53  kreator

	* ChangeLog: Added Changelog at last. Using cvs2cl.

2001-03-14 23:42  kreator

	* sample/DEBIAN/: conffiles, control, postinst, prerm: samples

2001-03-14 23:42  kreator

	* sample/pam.conf: pam.conf

2001-03-14 23:42  kreator

	* Makefile.in, RELEASENOTES, TODO: TODO, Makefile, release..

2001-03-09 17:34  kreator

	* pam_skey.c, pam_skey_access.c: Minor changes - mod_opt ==
	_MOD_NONE

2001-03-08 23:56  kreator

	* pam_skey.c: Retab pam_skey.c

2001-03-08 12:11  kreator

	* README: Readme was bad looking.

2001-03-08 11:22  kreator

	* Makefile.in: Added linting.

2001-03-08 11:10  kreator

	* INSTALL, RELEASENOTES, pam_skey.c, pam_skey.h: Documentation
	update, degarbage() macro renamed into _pam_degarbage()

2001-03-08 01:47  kreator

	* RELEASENOTES, RELEASENOTES, autoconf/configure.in, defs.h.in:
	Misc.

2001-03-08 01:44  kreator

	* configure, defs.h.in, misc.h.in, pam_skey.c, pam_skey_access.c,
	autoconf/acconfig.h, autoconf/configure.in: Misc changes

2001-03-08 01:20  kreator

	* Makefile.in, configure, defs.h.in, misc.h.in, pam_skey.c,
	pam_skey_access.c, autoconf/config.guess, autoconf/config.sub,
	autoconf/configure.in: Lots of misc changes.

2001-03-08 00:02  kreator

	* RELEASENOTES, pam_skey.c, pam_skey.h: 1.1.0, solaris2.7 puts '\n'
	at the end of string - this fixes it.

2001-03-07 21:29  kreator

	* INSTALL, README: Misc documentation update.

2001-03-07 20:51  kreator

	* Makefile.in, RELEASENOTES, pam_skey.c, pam_skey.h,
	pam_skey_access.c: file skeylogin.c function int skeyinfo() <...> 
	case 1:   /* User not found */	fclose(mp->keyfile);	return -1; 
			   <-------- obvious	    wrong return value <..>

2001-03-05 07:58  kreator

	* configure: I forgot to update configure in root

2001-03-05 07:57  kreator

	* Makefile.in, README, RELEASENOTES, autoconf/configure.in: 1.1.0 -
	Rules for Makefile.in, README updates, etc.

2001-03-04 23:09  kreator

	* INSTALL: Documentation changes - INSTALL

2001-03-04 19:18  kreator

	* RELEASENOTES, pam_skey.h: Release.

2001-03-04 19:05  kreator

	* RELEASENOTES: Added releasenotes. Ready to publish.

2001-03-04 18:43  kreator

	* README: Credits - go out! :)

2001-03-04 15:19  kreator

	* Makefile.in, pam_skey.h: -pendantic in Makefile redo option
	redefinitions in pam_skey.h

2001-03-04 15:04  kreator

	* README: README changes.

2001-03-04 14:33  kreator

	* INSTALL, README: Minor changes in INSTALL and README files.

2001-03-03 23:24  kreator

	* INSTALL, Makefile.in, README: Smaller set of changes to get $Id: ChangeLog,v 1.1.1.1 2005/06/18 12:11:24 kreator Exp $
	working.

2001-03-03 23:22  kreator

	* README: Smaller set of changes in README files. Credits,
	information, etc.

2001-03-03 23:12  kreator

	* autoconf/configure.in: Add $Id: ChangeLog,v 1.1.1.1 2005/06/18 12:11:24 kreator Exp $ to configure.in Perhaps should
	make serial?

2001-03-03 23:11  kreator

	* configure, autoconf/configure.in: Small changes to configure.
	Start talking in english, this croatian is PITA :)

2001-03-03 22:56  kreator

	* INSTALL: Dotjerivanje INSTALL-a

2001-03-03 22:46  kreator

	* COPYING, INSTALL, Makefile.in, README, configure, defs.h.in,
	autoconf/acconfig.h, autoconf/configure.in, autoconf/install-sh,
	autoconf/stamp-h.in: Initial revision

2001-03-03 22:46  kreator

	* COPYING, INSTALL, Makefile.in, README, configure, defs.h.in,
	autoconf/acconfig.h, autoconf/configure.in, autoconf/install-sh,
	autoconf/stamp-h.in: Imported from RCS versions

2001-03-03 20:47  kreator

	* pam_skey.c: Zaboravio \n kod fprintf(stderr, "s/key database
	error...

2001-03-01 23:47  kreator

	* pam_skey_access.c: tab

2001-03-01 22:34  kreator

	* pam_skey.c, pam_skey.h: Dodana opcija _MOD_ONLY_ONE_TRY

2001-03-01 21:10  kreator

	* pam_skey.c: Potkrao se ozbiljan bug, radio sam memset(.., 0, ...)
	nakon popunjavanja strukture u proslom releasu. Odmah ispravljeno..

2001-03-01 19:36  kreator

	* pam_skey.c, pam_skey_access.c: Radije da bude compile clean za
	Solaris

2001-03-01 17:48  kreator

	* pam_skey_access.c: Ekstra poliranje ulaznih i izlaznih parametara

2001-03-01 17:47  kreator

	* pam_skey.c: Ekstra ciscenje i poliranje ulaznih i izlaznih
	argumenata i njihovih tipova

2001-03-01 17:34  kreator

	* pam_skey.c, pam_skey_access.c: PAM_EXTERN u source prije glavnog
	pam.h

2001-03-01 17:32  kreator

	* pam_skey.h: Izbacio PAM_EXTERN, mora ici u source prije
	pamblabla.h

2001-03-01 17:24  kreator

	* pam_skey.h: Fix za empty_authtok

2001-03-01 01:53  kreator

	* pam_skey.c, pam_skey.h: malo pociscen kod, dodan makro
	empty_authtok

2001-03-01 01:46  kreator

	* pam_skey.c, pam_skey_access.c: char ** cast za pam_get_user

2001-02-28 01:04  kreator

	* pam_skey.c, pam_skey.h, pam_skey_access.c: Sitnice

2001-02-27 18:23  kreator

	* pam_skey_access.c: Initial revision

2001-02-26 23:16  kreator

	* pam_skey.c, pam_skey.h: unsigned mod_opt

2001-02-26 23:13  kreator

	* pam_skey.c: Kako sam promijenio mod_opt na uint, onda je trebalo
	i funkciju

2001-02-26 23:13  kreator

	* pam_skey.h: Kako sam promijenio mod_opt na uint, onda je trebalo
	i prototipove

2001-02-26 22:59  kreator

	* pam_skey.h: Makroi potrebni po defaultu i pod Linuxom


pam_skey.c
/* 
 * (c) 2001 Dinko Korunic, kreator@srce.hr
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *
 * S/KEY is a trademark of Bellcore.
 * Mink is the former name of the S/KEY authentication system.
 *
 * Programs that had some influence in development of this source:
 *  Wietse Venema's logdaemon package
 *  Olaf Kirch's Linux S/Key package
 *  Linux-PAM modules and templates
 *  Wyman Miles' pam_securid module
 *
 * Should you choose to use and/or modify this source code, please do so
 * under the terms of the GNU General Public License under which this
 * program is distributed.
 */

static char rcsid[] = "$Id: pam_skey.c,v 1.2 2005/06/18 12:36:18 kreator Exp $";

#include "defs.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef STRING_WITH_STRINGS
# include <strings.h>
#endif
#include <unistd.h>
#include <pwd.h> 
#include <sys/types.h>
#include <syslog.h>

#define PAM_EXTERN extern
#undef PAM_STATIC

#include <security/pam_appl.h>
#include <security/pam_modules.h>

#include "skey.h"
#include "pam_skey.h"
#include "misc.h"

PAM_EXTERN int pam_sm_setcred (pam_handle_t *pamh, int flags,
  int argc, const char **argv)
{
  return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags,
  int argc, const char **argv)
{
  char challenge[CHALLENGE_MAXSIZE]; /* challenge to print in conv */
  char msg_text[PAM_MAX_MSG_SIZE]; /* text for pam conv */
  char *username = NULL; /* username spacer */
  char *response = NULL; /* response spacer */
  struct skey skey; /* structure that contains skey information */
  int status; /* return status spacer */
  unsigned mod_opt = _MOD_NONE_ON; /* module options */

  /* Get module options */
  mod_getopt(&mod_opt, argc, argv);

  /* Get username */
#if defined LINUX || defined BSD
  if (pam_get_user(pamh, (const char **)&username, "login:")
#else
  if (pam_get_user(pamh, (char **)&username, "login:")
#endif
      != PAM_SUCCESS)
  {
    fprintf(stderr, "cannot determine username\n");
    if (mod_opt & _MOD_DEBUG)
      syslog(LOG_DEBUG, "cannot determine username");
    return PAM_USER_UNKNOWN;
  }

  if (mod_opt & _MOD_DEBUG)
    syslog(LOG_DEBUG, "got username %s", username);

#ifdef HAVE_SKEYACCESS
  /* Check S/Key access permissions - user, host and port. Also include
   * sanity checks */
  if (mod_opt & _MOD_ACCESS_CHECK)
  {
    char *host; /* points to host */
    char *port; /* points to port */
    struct passwd *pwuser; /* structure for getpw() */

    /* Get host.. */
#if defined LINUX || defined BSD
    if (pam_get_item(pamh, PAM_RHOST, (const void **)&host)
#else
    if (pam_get_item(pamh, PAM_RHOST, (void **)&host)
#endif
        != PAM_SUCCESS)
      host = NULL; /* couldn't get host */
    /* ..and port */
#if defined LINUX || defined BSD
    if (pam_get_item(pamh, PAM_TTY, (const void **)&port)
#else
    if (pam_get_item(pamh, PAM_TTY, (void **)&port)
#endif
        != PAM_SUCCESS)
      port = NULL; /* couldn't get port */

    if (mod_opt & _MOD_DEBUG)
      syslog(LOG_DEBUG, "checking s/key access for user %s,"
        " host %s, port %s", username,
        (host != NULL) ? host : "*unknown*",
        (port != NULL) ? port : "*unknown*");

    /* Get information from passwd file */
    if ((pwuser = getpwnam(username)) == NULL)
    {
      fprintf(stderr, "no such user\n");
      syslog(LOG_NOTICE, "cannot find user %s", username);
      return PAM_USER_UNKNOWN; /* perhaps even return PAM_ABORT here? */
    }

    /* Do actual checking - we assume skeyaccess() returns PERMIT which is
     * by default 1. Notice 4th argument is NULL - we will not perform
     * address checks on host itself */
    if (skeyaccess(pwuser, port, host, NULL) != 1)
    {
      fprintf(stderr, "no s/key access permissions\n");
      syslog(LOG_NOTICE, "no s/key access permissions for %s",
          username);
      return PAM_AUTH_ERR;
    }
  }
  else

#endif /* HAVE_SKEYACCESS */
    
  /* Only do check whether user has passwd entry */
    if (getpwnam(username) == NULL)
    {
      fprintf(stderr, "no such user\n");
      if (mod_opt & _MOD_DEBUG)
        syslog(LOG_DEBUG, "cannot find user %s",
            username);
      return PAM_USER_UNKNOWN;
    }

  /* Get S/Key information on user with skeyinfo() */
#ifdef HAVE_SKEYINFO
  switch (skeyinfo(&skey, username, NULL))
#else
#ifdef HAVE_SKEYLOOKUP
  switch (skeylookup(&skey, username))
#endif /* HAVE_SKEYLOOKUP */
#endif /* HAVE_SKEYINFO */
  {
  /* 0: OK */
  case 0:
    break;
  /* -1: File error */
  case -1:
#if 0
  /* XXX- This seems broken in (at least) logdaemon-5.8. It returns -1
   * when user not found in keyfile. -kre */
    fprintf(stderr, "s/key database error\n");
    syslog(LOG_NOTICE, "s/key database error");
    return PAM_AUTH_ERR;
#endif
  /* 1: No such user in database */
  case 1:
    /* We won't confuse the ordinary user telling him about missing skeys
     * -kre */
#if 0
    fprintf(stderr, "no s/key for %s\n", username);
#endif
    if (mod_opt & _MOD_DEBUG)
      syslog(LOG_DEBUG, "no s/key for %s\n", username);
    return PAM_AUTH_ERR;
  }

  /* Make challenge string */
#if defined(SKEY_MAX_HASHNAME_LEN) && defined(SKEY_MAX_SEED_LEN)
  snprintf(challenge, CHALLENGE_MAXSIZE, "otp-%.*s %d %.*s",
      SKEY_MAX_HASHNAME_LEN, skey_get_algorithm(), skey.n - 1, SKEY_MAX_SEED_LEN, skey.seed);
#else
  snprintf(challenge, CHALLENGE_MAXSIZE, "s/key %d %s",
      skey.n - 1, skey.seed);
#endif

  if (mod_opt & _MOD_DEBUG)
    syslog(LOG_DEBUG, "got challenge %s for %s", challenge,
        username);

  /* Read response from last module's PAM_AUTHTOK */
  if (mod_opt & _MOD_USE_FIRST_PASS)
  {
    /* Try to extract authtoken */
#if defined LINUX || defined BSD
    if (pam_get_item(pamh, PAM_AUTHTOK, (const void **)&response)
#else
    if (pam_get_item(pamh, PAM_AUTHTOK, (void **)&response)
#endif
        != PAM_SUCCESS)
    {
      if (mod_opt & _MOD_DEBUG)
        syslog(LOG_DEBUG, "could not get PAM_AUTHTOK");
      mod_opt &= ~_MOD_USE_FIRST_PASS;
    }
    else
    {
      /* Got AUTHTOK, but it was empty */
      if (empty_authtok(response))
      {
        if (mod_opt & _MOD_DEBUG)
          syslog(LOG_DEBUG, "empty PAM_AUTHTOK");
        mod_opt &= ~_MOD_USE_FIRST_PASS;
      }
      else
        /* All OK, print challenge information */
        fprintf(stderr, "challenge %s\n", challenge);
    }
  }

  /* There was no PAM_AUTHTOK, or there was no such option in pam-conf
   * file */
  if (!(mod_opt & _MOD_USE_FIRST_PASS))
  {
    /* Prepare a complete message for conversation */
    snprintf(msg_text, PAM_MAX_MSG_SIZE,
        "challenge %s\npassword: ", challenge);

    /* Talk with user */
    if (mod_talk_touser(pamh, &mod_opt, msg_text, &response)
        != PAM_SUCCESS)
      return PAM_SERVICE_ERR;

    /* Simulate standard S/Key login procedure - if empty token, turn on
     * ECHO and prompt again */
    if (empty_authtok(response) && !(mod_opt & _MOD_ONLY_ONE_TRY))
    {
      /* Was there echo off? */
      if (mod_opt & _MOD_ECHO_OFF)
      {
        _pam_delete(response);
        fprintf(stderr, "(turning echo on)\n");
        mod_opt &= ~_MOD_ECHO_OFF;

        /* Prepare a complete message for conversation */
        snprintf(msg_text, PAM_MAX_MSG_SIZE, "password: ");

        /* Talk with user */
        if (mod_talk_touser(pamh, &mod_opt, msg_text, &response)
          != PAM_SUCCESS)
          return PAM_SERVICE_ERR;

        /* Got again empty response. Bailout and don't save auth token */
        if (empty_authtok(response))
          return PAM_AUTH_ERR;
      }
      else
      /* There was echo on already - just get out and don't save auth token
       * for other modules */
        return PAM_AUTH_ERR;
    }

    /* XXX - ECHO ON puts '\n' at the end in Solaris 2.7! This is
     * cludge to get rid of this nasty `feature' -kre */
    _pam_degarbage(response);
  
    /* Store auth token - that next module can use with `use_first_pass' */
    if (pam_set_item(pamh, PAM_AUTHTOK, response) != PAM_SUCCESS)
    {
      syslog(LOG_NOTICE, "unable to save auth token");
      return PAM_SERVICE_ERR;
    }

    /* cleanup conversation */
    _pam_delete(response);
  } 

  /* Verify S/Key */
  status = skeyverify(&skey, response);

  switch (status)
  {
    /* 0: Verify successful, database updated */
    case 0:
      break;
    /* -1: Error of some sort; database unchanged */
    /*  1: Verify failed, database unchanged */
    case -1:
    case 1:
      if (mod_opt & _MOD_DEBUG)
        syslog(LOG_DEBUG, "verify for %s failed, database"
            " unchanged", username);
      return PAM_AUTH_ERR;
  }

  /* Success by default */
  return PAM_SUCCESS;
}

/* Get module optional parameters */
static void mod_getopt(unsigned *mod_opt, int mod_argc, const char **mod_argv)
{
  int i;

  /* Setup runtime defaults */
  *mod_opt |= _MOD_DEFAULT_FLAG;
  *mod_opt &= _MOD_DEFAULT_MASK;

  /* Setup runtime options */
  while (mod_argc--)
  {
    for (i = 0; i < _MOD_ARGS; ++i)
    {
      if (mod_args[i].token != NULL &&
          !strncmp(*mod_argv, mod_args[i].token,
            strlen(mod_args[i].token)))
        break;
    }
    if (i >= _MOD_ARGS)
      syslog(LOG_ERR, "unknown option %s", *mod_argv);
    else
    {
      *mod_opt &= mod_args[i].mask; /* Turn off */
      *mod_opt |= mod_args[i].flag; /* Turn on */
    }
    ++mod_argv;
  }
}

/* This will talk to user through PAM_CONV */
static int mod_talk_touser(pam_handle_t *pamh, unsigned *mod_opt,
    char *msg_text, char **response)
{
  struct pam_message message;
  const struct pam_message *pmessage = &message;
  struct pam_conv *conv = NULL;
  struct pam_response *presponse = NULL;

  /* Better safe than sorry */
  *response = NULL;

  /* Be paranoid */
  memset(&message, 0, sizeof(message));

  /* Turn on/off PAM echo */
  if (*mod_opt & _MOD_ECHO_OFF)
    message.msg_style = PAM_PROMPT_ECHO_OFF;
  else
    message.msg_style = PAM_PROMPT_ECHO_ON;
  
  /* Point to conversation text */
  message.msg = msg_text;

  /* Do conversation and see if all is OK */
#if defined LINUX || defined BSD
  if (pam_get_item(pamh, PAM_CONV, (const void **)&conv)
#else
  if (pam_get_item(pamh, PAM_CONV, (void **)&conv)
#endif
      != PAM_SUCCESS)
  {
    if (*mod_opt & _MOD_DEBUG)
      syslog(LOG_DEBUG, "error in conversation");
    return PAM_SERVICE_ERR;
  }

  /* Convert into pam_response - only 1 reply expected */
#if defined LINUX || defined BSD
  if (conv->conv(1, &pmessage, &presponse,
        conv->appdata_ptr)
#else
  if (conv->conv(1, (struct pam_message **)&pmessage, &presponse,
        conv->appdata_ptr)
#endif
    != PAM_SUCCESS)
  {
    _pam_delete(presponse->resp);
    return PAM_SERVICE_ERR;
  }

  if (presponse != NULL)
  {
    /* Save address */
    *response = presponse->resp;
    /* To ensure that response address will not be erased */
    presponse->resp = NULL;
    _pam_drop(presponse);
  }
  else
    return PAM_SERVICE_ERR;

  return PAM_SUCCESS;
}
